serverless/azure/src/index.ts

```diff
@@ -0,0 +1,81 @@
+import {
+  app,
+  HttpRequest,
+  HttpResponseInit,
+  InvocationContext,
+} from "@azure/functions";
+import { Headers } from "undici";
+import { getZxy } from "./get";
+import { FetchSource, Source } from "pmtiles";
+import { tile_path } from "./tile_path";
+import { getAzureStorageSource } from "./azure_source";
+
+function getSource(name: string): Source | null {
+  let mapUrl = process.env["PMTILES_PATH"];
+  const mapName = name ?? "default";
+
+  if (mapName !== "default") {
+    mapUrl = process.env[`PMTILES_PATH_${name}`];
+  }
+
+  if (mapUrl) {
+    return new FetchSource(mapUrl);
+  }
+
+  return getAzureStorageSource(mapName);
+}
+
+export async function httpTrigger(
+  request: HttpRequest,
+  context: InvocationContext
+): Promise<HttpResponseInit> {
+  const url = new URL(request.url);
+  const { ok, name, tile, ext } = tile_path(
+    url.pathname,
+    process.env["TILE_PATH"]
+  );
+
+  let allowed_origin = "*";
+  if (typeof process.env["ALLOWED_ORIGINS"] !== "undefined") {
```

Comment:
This might not interact with a caching CDN in front correctly - if we store the `Access-Control-Allow-Origin` in the CDN cache, we need to invalidate the entire cache any time we change the set of allowed hosts. For AWS we can work around this by letting CloudFront dynamically handle CORS, for Cloudflare the Workers code can determine the correct header on every invocation. Is there any Azure CDN functionality to do dynamic CORS like CloudFront?

Reply:
Hmm, I need to check that. I used `*` in our case. I pretty sure Azure CDN can set the header dynamically.

Also, I see that the `ALLOWED_ORIGINS` parameter is used in a different place as well.

Reply:
I asked ChatGDP. Got this answer:

1. Vary Header: If your service correctly implements CORS, it should include a Vary: Origin header in its responses. This header tells caches, including CDNs, that the response varies based on the value of the Origin request header. When a CDN sees this, it should cache separate copies of the resource for each unique value of the Origin header. In this case, a request from example.com would hit one cache entry, and a request from other.com would hit a different cache entry, assuming both are allowed origins.

1. Caching Without Vary Header: If your service does not include a Vary: Origin header, the CDN might cache the response to the first request and then serve that cached response to the second request, regardless of the origin. This can lead to incorrect Access-Control-Allow-Origin headers being sent to clients and potential CORS errors in the browser.

1. CDN Configuration: Some CDNs offer more advanced configurations that can help manage CORS headers more effectively. For example, you might be able to configure the CDN to dynamically alter the Access-Control-Allow-Origin header based on the origin of the request. However, this capability depends on the specific CDN and its feature set.

1. Potential Issues: Without proper handling, you might encounter issues where a user from an unallowed origin receives a cached response indicating they are allowed, or a user from an allowed origin receives a response indicating they are not allowed, due to the CDN serving a cached response intended for a different origin.

Testing and Verification: It's crucial to test how your CDN handles requests from different origins, especially if you are dynamically setting the Access-Control-Allow-Origin header. Verify that the correct headers are returned and that CORS is functioning as expected for users from different domains.

Reply:
In summary, I think that in this case it is probably better to have the CDN handle CORS. That way the tile will be cached in the CDN, based on URL and cache-control headers, and subsequent requests will use the cached entry (I assume).